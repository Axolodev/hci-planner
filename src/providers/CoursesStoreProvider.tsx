import {
  CoursesStoreType,
  createCoursesStore,
  initCounterStore,
} from "@/stores/coursesStore";
import { createContext, ReactNode, useContext, useRef } from "react";
import { useStore } from "zustand";

// Define provides for React
export type CoursesStoreApi = ReturnType<typeof createCoursesStore>;

export const CoursesStoreContext = createContext<CoursesStoreApi | undefined>(
  undefined
);

export interface CoursesStoreProviderProps {
  children: ReactNode;
}

export const CoursesStoreProvider = ({
  children,
}: CoursesStoreProviderProps) => {
  const storeRef = useRef<CoursesStoreApi | null>(null);
  if (storeRef.current === null) {
    storeRef.current = createCoursesStore(initCounterStore());
  }

  return (
    <CoursesStoreContext.Provider value={storeRef.current}>
      {children}
    </CoursesStoreContext.Provider>
  );
};

// Define a hook to access the store
export const useCoursesStore = <T,>(
  selector: (store: CoursesStoreType) => T
): T => {
  const coursesStoreContext = useContext(CoursesStoreContext);

  if (!coursesStoreContext) {
    throw new Error(`useCoursesStore must be used within CoursesStoreProvider`);
  }

  return useStore(coursesStoreContext, selector);
};
