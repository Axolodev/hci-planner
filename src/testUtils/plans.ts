import { Plan } from "@/types";

export const basicPlan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};

export const planWithMultipleSectionOptions: Plan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
        {
          name: "Another option",
          modules: [
            {
              name: "Another module",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};
