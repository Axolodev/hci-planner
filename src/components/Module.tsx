import { useCoursesStore } from "@/providers/CoursesStoreProvider";
import { Module as ModuleType } from "@/types";
import * as React from "react";

export const TESTING_ID = "Module";

interface Props extends ModuleType {
  moduleIndex: number;
  optionName: string;
  sectionTitle: string;
  optionIndex: number;
  disabled: boolean;
}

const Module: React.FunctionComponent<Props> = ({
  name,
  sectionTitle,
  optionIndex,
  moduleIndex,
  optionName,
  disabled,
}) => {
  const { getModuleStatus, setModuleStatus } = useCoursesStore(
    (state) => state
  );
  const isCompleted = getModuleStatus(sectionTitle, optionIndex, moduleIndex);
  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    setModuleStatus(
      sectionTitle,
      optionIndex,
      moduleIndex,
      event.target.checked
    );
  };

  return (
    <input
      type="checkbox"
      onChange={handleChange}
      checked={isCompleted}
      disabled={disabled}
      id={`${optionName}-${name}`}
      className={`btn join-item disabled:cursor-not-allowed w-full rounded-field min-h-[40px] h-auto`}
      aria-label={name}
      style={{ margin: 0 }}
      data-testid={TESTING_ID}
    />
  );
};

export default Module;
