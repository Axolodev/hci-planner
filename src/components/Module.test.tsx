import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CoursesStoreProvider } from "../providers/CoursesStoreProvider";
import Module, { TESTING_ID } from "./Module";
import { basicPlan } from "../testUtils/plans";

describe("Module", () => {
  it.skip("renders correctly and matches snapshot", () => {
    const { container } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <Module
          sectionTitle="Section title"
          optionIndex={0}
          moduleIndex={0}
          optionName="Option name"
          disabled={false}
        />
      </CoursesStoreProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("can be checked", async () => {
    const { getByTestId } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <Module
          optionIndex={0}
          moduleIndex={0}
          optionName="Option name"
          disabled={false}
          sectionTitle="Section title"
        />
      </CoursesStoreProvider>
    );
    const checkbox = getByTestId(TESTING_ID) as HTMLInputElement;
    expect(checkbox.checked).toBe(false);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);
  });
});
