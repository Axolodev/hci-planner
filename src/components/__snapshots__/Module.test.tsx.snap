// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Module renders correctly and matches snapshot 1`] = `
<div>
  <input
    class="btn join-item disabled:cursor-not-allowed w-full rounded-field"
    data-testid="Module"
    id="Option name-undefined"
    style="margin: 0px;"
    type="checkbox"
  />
</div>
`;
